src/hypertable_restrict_info.c

```diff
@@ -71,68 +73,128 @@ dimension_restrict_info_create(Dimension *d)
 {
-	int64		value = time_value_to_internal(c->constvalue, c->consttype, false);
+	if (!use_or && list_length(dri->partitions) > 1)
+		/* since this is AND the intersection is empty set */
+	{
+		dri->partitions = NIL;
+		return true;
+	}
+	return false;
+}
 
-	switch (strategy)
+static bool
+dimension_restrict_info_open_add(DimensionRestrictInfoOpen *dri, StrategyNumber strategy, List *values, Oid type)
+{
+	ListCell   *item;
+	bool		restriction_added = false;
+
+	foreach(item, values)
 	{
-		case BTLessEqualStrategyNumber:
-		case BTLessStrategyNumber:
-			if (dri->upper_strategy == InvalidStrategy || value < dri->upper_bound)
-			{
-				dri->upper_strategy = strategy;
-				dri->upper_bound = value;
-			}
-			return true;
-		case BTGreaterEqualStrategyNumber:
-		case BTGreaterStrategyNumber:
-			if (dri->lower_strategy == InvalidStrategy || value > dri->lower_bound)
-			{
-				dri->lower_strategy = strategy;
+		Datum		datum = PointerGetDatum(lfirst(item));
+		int64		value = time_value_to_internal(datum, type, true);
+
+		if (value == -1)
+		{
+			/* open dimension does not support IN/ANY/ALL restrictions */
+			elog(WARNING, "unknown time restriction type %d", type);
+			return restriction_added;
+		}
+
+		switch (strategy)
+		{
+			case BTLessEqualStrategyNumber:
+			case BTLessStrategyNumber:
+				if (dri->upper_strategy == InvalidStrategy || value < dri->upper_bound)
+				{
+					dri->upper_strategy = strategy;
+					dri->upper_bound = value;
+					restriction_added = true;
+				}
+				break;
+			case BTGreaterEqualStrategyNumber:
+			case BTGreaterStrategyNumber:
+				if (dri->lower_strategy == InvalidStrategy || value > dri->lower_bound)
+				{
+					dri->lower_strategy = strategy;
+					dri->lower_bound = value;
+					restriction_added = true;
+				}
+				break;
+			case BTEqualStrategyNumber:
 				dri->lower_bound = value;
-			}
-			return true;
-		case BTEqualStrategyNumber:
-			dri->lower_bound = value;
-			dri->upper_bound = value;
-			dri->lower_strategy = BTGreaterEqualStrategyNumber;
-			dri->upper_strategy = BTLessEqualStrategyNumber;
-			return true;
-		default:
-			return false;
+				dri->upper_bound = value;
+				dri->lower_strategy = BTGreaterEqualStrategyNumber;
+				dri->upper_strategy = BTLessEqualStrategyNumber;
+				restriction_added = true;
+				break;
+			default:
+				/* unsupported strategy */
+				break;
+
+		}
 	}
+	return restriction_added;
 }
 
 static bool
-dimension_restrict_info_closed_add(DimensionRestrictInfoClosed *dri, StrategyNumber strategy, Const *c)
+dimension_restrict_info_closed_add(DimensionRestrictInfoClosed *dri, StrategyNumber strategy, List *values, bool use_or)
 {
-	int64		value = partitioning_func_apply(dri->base.dimension->partitioning, c->constvalue);
+	ListCell   *item;
+	bool		restriction_added = false;
 
-	switch (strategy)
+	if (strategy != BTEqualStrategyNumber)
 	{
-		case BTEqualStrategyNumber:
-			dri->value = value;
+		return restriction_added;
+	}
+
+	if (dri->partitions == NIL)
```

Comment:
I think this is the wrong check dri->partitions can be NIL in 2 cases:
1) dri->partitions was never set because the dri was just created.
2) all partitions were excluded (the list is empty)
I think it's incorrect to conflate the two.

 I think the test you want here is dri->strategy == InvalidStrategy